hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java

```diff
@@ -110,8 +129,126 @@ public SqlResult execute(@Nonnull SqlStatement statement) {
+                resubmissionResult = createResubmissionResult(message, connection);
+                if (resubmissionResult.getSqlError() == null) {
+                    logFinest(logger, "Resubmitting query: %s ended without error", result.getQueryId());
+                } else {
+                    logFinest(logger, "Resubmitting query: %s ended with error", result.getQueryId());
+                }
+                if (resubmissionResult.getSqlError() == null || !shouldResubmit(resubmissionResult.getSqlError())) {
+                    return resubmissionResult;
+                }
+            } catch (Exception e) {
+                logFinest(logger, "Resubmitting query: %s ended with exception", result.getQueryId());
+                RuntimeException rethrown = connection == null ? (RuntimeException) e : rethrow(e, connection);
+                if (!shouldResubmit(rethrown)) {
+                    throw rethrown;
+                }
+            }
+
+            // We measure the retry count and retry timeout for each individual resubmission. If a resubmission succeeds,
+            // and then later some fetch of the same query fails, we'll try to resubmit again, and we'll not include
+            // the number and time spent in the previous resubmission.
+            if (invokeCount++ >= MAX_FAST_INVOCATION_COUNT) {
+                long delayMillis =
+                        Math.min(1L << Math.min(62, invokeCount - MAX_FAST_INVOCATION_COUNT), resubmissionRetryPauseMillis);
+                try {
+                    Thread.sleep(delayMillis);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    return returnNonNullOrThrow(resubmissionResult, error, true);
+                }
+            }
+        } while (System.nanoTime() - resubmissionStartTime <= resubmissionTimeoutNano);
+
+        logger.finest("Resubmitting query timed out");
+
+        return returnNonNullOrThrow(resubmissionResult, error, false);
+    }
+
+    private SqlResubmissionResult returnNonNullOrThrow(
+            SqlResubmissionResult result,
+            RuntimeException originalError,
+            boolean wasInterrupted
+    ) {
+        if (result == null) {
+            // We have nothing to return to the caller, no valid result and no new exception, so we throw the original error
+            // packed in proper exception so the caller has valid information what happened.
+            if (wasInterrupted) {
+                throw new HazelcastException("Query resubmission was interrupted", originalError);
+            }
+            throw new OperationTimeoutException("Query resubmission timed out", originalError);
+        }
+        return result;
+    }
+
+    private boolean shouldResubmit(Exception error) {
+        return (error instanceof HazelcastSqlException) && (shouldResubmit(((HazelcastSqlException) error).getCode()));
```

Comment:
I think İhsan was talking about this. If we get `TargetDisconnectedException` here, because the connection you send the execute query is closed, should we handle that?

Reply:
@mdumandag We are getting connection from ```client.getConnectionManager().getConnectionForSql()``` where ```client``` is an instance of ```HazelcastClientInstanceImpl```. Can that return closed connection? Or can it close during invocation and throw that exception?

Reply:
It won't return a closed connection, but the connection can be closed right after you get it. It can also close while doing the invocation, or while waiting for the invocation response